lib/smb2/file.rb

```diff
@@ -84,34 +105,59 @@ def read(offset: 0, length: nil)
   # @param length [Fixnum] number of bytes to read, starting from `offset`. If
   #   this is greater than the server's maximum read size, the server will
   #   respond with STATUS_INVALID_PARAMETER
-  # @return [String]
-  def read_chunk(offset: 0, length: self.tree.client.max_read_size)
-    packet = Smb2::Packet::ReadRequest.new do |request|
-      request.read_offset = offset
-      request.read_length = length
-      request.file_id = self.create_response.file_id
-      request.minimum_count = 0
-    end
+  # @return [Smb2::Packet::ReadResponse]
+  def read_chunk(offset: self.pos, length: self.tree.client.max_read_size)
+    packet = Smb2::Packet::ReadRequest.new(
+      read_offset: offset,
+      read_length: length,
+      file_id: self.create_response.file_id,
+      minimum_count: 0
+    )
 
     response = tree.send_recv(packet)
 
     response_packet = Smb2::Packet::ReadResponse.new(response)
     @last_read_response = response_packet
 
-    response_packet.data
+    seek(response_packet.data_length, IO::SEEK_CUR)
+
+    response_packet
+  end
+
+  # Seeks to a given `offset` in the stream according to the value of
+  # `whence`:
+  #
+  # |`whence`||
+  # |----------------------|--------------------------------------------------
+  # | :CUR or IO::SEEK_CUR | Seeks to `amount` plus current position
+  # | :END or IO::SEEK_END | Seeks to `amount` plus end of stream (you probably want a negative value for `amount`)
+  # | :SET or IO::SEEK_SET | Seeks to the absolute location given by `amount`
+  #
+  # @see http://ruby-doc.org/core-2.2.2/IO.html#method-i-seek
+  # @param amount [Fixnum]
+  # @param whence [Symbol]
+  def seek(amount, whence=IO::SEEK_SET)
+    @pos = case whence
+           when :CUR, IO::SEEK_CUR
+             @pos + amount
+           when :END, IO::SEEK_END
+             amount + size
+           when :SET, IO::SEEK_SET
+             amount
+           end
   end
 
   # The size of the file in bytes
   #
   # @return [Fixnum]
   def size
-    packet = Smb2::Packet::QueryInfoRequest.new do |request|
-      request.info_type = Smb2::Packet::QUERY_INFO_TYPES[:FILE]
-      request.file_info_class = Smb2::Packet::FILE_INFORMATION_CLASSES[:FileStandardInformation]
-      request.output_buffer_length = 40
-      request.input_buffer_length = 0
-      request.file_id = self.create_response.file_id
-    end
+    packet = Smb2::Packet::QueryInfoRequest.new(
+      info_type: Smb2::Packet::QUERY_INFO_TYPES[:FILE],
+      file_info_class: Smb2::Packet::FILE_INFORMATION_CLASSES[:FileStandardInformation],
+      output_buffer_length: 40,
```

Comment:
Why `40`?  Can you put this in a constant with a descriptive comment explain why?